core/src/main/scala/besom/types.scala

```diff
@@ -157,21 +163,61 @@ object types:
   object URN:
     /** The instance of [[URN]] that represents an empty URN
       */
-    val empty: URN = ""
+    private[besom] val empty: URN = ""
 
-    // This is implemented according to https://www.pulumi.com/docs/concepts/resources/names/#urns
-    private[types] val UrnRegex =
-      """urn:pulumi:(?<stack>[^:]+)::(?<project>[^:]+)::(?<parentType>.+?)((\$(?<resourceType>.+))?)::(?<resourceName>.+)""".r
+    /** This is implemented according to https://www.pulumi.com/docs/concepts/resources/names/#urns and
+      * https://pulumi-developer-docs.readthedocs.io/en/latest/providers/implementers-guide.html?highlight=URN#urns
+      *
+      * If you want to understand this regex better head to: https://regex101.com/r/o2QWJ3/1
+      *
+      * Now hear me out, this regex is a bit complicated, but it's not that bad. Let's break it down. We have to adhere to this grammar:
```

Comment:
I like your humor, "not that bad regex" is dark, but still funny ;)

Don't worry, one day we'll write a proper parser :)